stellargraph/layer/graph_attention.py

```diff
@@ -0,0 +1,481 @@
+
+            # Apply dropout to features and attention coefficients
+            dropout_feat = Dropout(self.in_dropout_rate)(features)  # (N x F')
+            dropout_attn = Dropout(self.attn_dropout_rate)(dense)  # (N x N)
+
+            # Linear combination with neighbors' features [YT: see Eq. 4]
+            node_features = K.dot(dropout_attn, dropout_feat)  # (N x F')
+
+            if self.use_bias:
+                node_features = K.bias_add(node_features, self.biases[head])
+
+            # Add output of attention head to final output
+            outputs.append(node_features)
+
+        # Aggregate the heads' output according to the reduction method
+        if self.attn_heads_reduction == "concat":
+            output = K.concatenate(outputs)  # (N x KF')
+        else:
+            output = K.mean(K.stack(outputs), axis=0)  # N x F')
+
+        output = self.activation(output)
+        return output
+
+    def compute_output_shape(self, input_shape):
+        output_shape = input_shape[0][0], self.output_dim
+        return output_shape
+
+
+class GAT:
+    """
+    A stack of GAT layers with aggregation of multiple attention heads, Eqs 5-6 of GAT paper
+    """
+
+    def __init__(
+        self,
+        layer_sizes,
+        attn_heads=1,
+        attn_heads_reduction=None,
+        activations=None,
+        bias=True,
+        in_dropout=0.0,
+        attn_dropout=0.0,
+        normalize="l2",
+        generator=None,
+    ):
+        """
+
+        Args:
+            layer_sizes: list of output sizes of GAT layers in the stack
+            attn_heads: number of attention heads
+            attn_heads_reduction:
+            activations: list of activations applied to each layer's output
+            bias: toggles an optional bias in GAT layers
+            in_dropout: dropout rate applied to input features of each GAT layer
+            attn_dropout: dropout rate applied to attention maps
+            normalize: normalization applied to the final output features of the GAT layers stack
+            generator: an instance of FullBatchNodeGenerator class constructed on the graph of interest
+        """
+        self._gat_layer = GraphAttention
+        self.attn_heads = attn_heads
+        self.bias = bias
+        self.in_dropout = in_dropout
+        self.attn_dropout = attn_dropout
+        self.generator = generator
+
+        if generator is not None:
+            assert isinstance(
+                generator, FullBatchNodeGenerator
+            ), "{}: generator must be of type FullBatchNodeGenerator or None; received object of type {} instead".format(
+                type(self).__name__, type(generator).__name__
+            )
+
+        if attn_heads_reduction is None:
+            # default head reductions, see eqs 5-6 of the GAT paper
+            self.attn_heads_reduction = ["concat"] * (len(layer_sizes) - 1) + [
+                "average"
+            ]
+        else:
```

Comment:
Perhaps a check that the length of `attn_heads_reduction` is correct would be a good idea.

Reply:
check is added